### PR TITLE
Add an option to allow rabbitmq_user to use API for managing users

### LINKED ISSUE #76: Add an option to allow rabbitmq_user to use API for managing users
##### SUMMARY
Currently rabbitmq_user uses `rabbitmqctl` for managing users. The idea is to add a choice for allowing to use API as well.

##### ISSUE TYPE

- Other plays in this collection already support using the API (like for managing exchanges, queues and bindings). My proposal is to add the option of using API for users as well without breaking backwards compatibility with the current implementation. 

The is to use providers to allow implementations to be kept and maintained apart from each other. I've done that before in the certifica role from linux system roles (https://github.com/linux-system-roles/certificate/). 

I would probably turn `module_utils/rabbitmq.py` in a package and add a directory providers inside it. The contents of rabbitmq.py could stay inside the package __init__.py.

##### ADDITIONAL INFORMATION

Using the 2 examples of the documentation that's how it would look:

```yaml
# Add user to server and assign full access control on / vhost.
# The user might have permission rules for other vhost but you don't care.
- community.rabbitmq.rabbitmq_user:
    user: joe
    password: changeme
    vhost: /
    configure_priv: .*
    read_priv: .*
    write_priv: .*
    state: present
    provider: api
    login_user: user
    login_password: secret
    login_host: remote.example.org

# Add user to server and assign full access control on / vhost.
# The user doesn't have permission rules for other vhosts
- community.rabbitmq.rabbitmq_user:
    user: joe
    password: changeme
    permissions:
      - vhost: /
        configure_priv: .*
        read_priv: .*
        write_priv: .*
    state: present
    provider: api
    login_user: user
    login_password: secret
    login_host: remote.example.org
 ```

### PR DESCRIPTION
**Pull request by rayhihi**
##### SUMMARY

As discussed in #76, this contains the solution we used in our company to manage rabbitmq users using the API.

Fixes #76

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
rabbitmq_user

##### ADDITIONAL INFORMATION 
With this feature is possible to create, delete and also manage your users. 
Unfortunately we won't be able to help with much more with the PR in the coming weeks so any help would be appreciated.

### CODE REVIEW

File: plugins/modules/rabbitmq_user.py [hunk 1]
```diff
@@ -106,6 +106,33 @@
     required: false
     default: on_create
     choices: ['on_create', 'always']
+  protocol:
+    description:
+      - Specify which TCP/IP protocol will be used
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:111]:
```suggestion
      - Specify which TCP/IP protocol will be used.
```

File: plugins/modules/rabbitmq_user.py [hunk 2]
```diff
@@ -106,6 +106,33 @@
...
+    type: str
+    default: http
+    choices: ['http', 'https']
+  host:
+    description:
+      - Hostname from API
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:117]:
```suggestion
      - Hostname from API.
```
I'm not sure, maybe to change `from` -> `of`?

File: plugins/modules/rabbitmq_user.py [hunk 3]
```diff
@@ -106,6 +106,33 @@
...
+    type: str
+    required: false
+  port:
+    description:
+      - Port of access from API
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:122]:
```suggestion
      - Port of access from API.
```

File: plugins/modules/rabbitmq_user.py [hunk 4]
```diff
@@ -106,6 +106,33 @@
...
+    type: str
+    required: false
+  login_user:
+    description:
+      - User from administrator of the management API
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:127]:
```suggestion
      - User from administrator of the management API.
```
Also maybe `User from administrator` -> `Administrator's username` or `Username with administrator access`?

File: plugins/modules/rabbitmq_user.py [hunk 5]
```diff
@@ -106,6 +106,33 @@
...
+    type: str
+    required: false
+  login_password:
+    description:
+      - Password from administrator of the management API
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:132]:
```suggestion
      - Login password of the management API.
```

File: plugins/modules/rabbitmq_user.py [hunk 6]
```diff
@@ -106,6 +106,33 @@
...
+    type: str
+    required: false
+
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:135]:
```suggestion
```

File: plugins/modules/rabbitmq_user.py [hunk 7]
```diff
@@ -106,6 +106,33 @@
     required: false
     default: on_create
     choices: ['on_create', 'always']
+  protocol:
+    description:
+      - Specify which TCP/IP protocol will be used
+    type: str
+    default: http
+    choices: ['http', 'https']
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:114]:
```suggestion
    choices: ['http', 'https']
    version_added: '1.2.0'
```

File: plugins/modules/rabbitmq_user.py [hunk 8]
```diff
@@ -106,6 +106,33 @@
...
+  host:
+    description:
+      - Hostname from API
+    type: str
+    required: false
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:119]:
```suggestion
    required: false
    version_added: '1.2.0'
```
also just a fyi required is false by default, so no need to specify it explicitly.

File: plugins/modules/rabbitmq_user.py [hunk 9]
```diff
@@ -106,6 +106,33 @@
...
+  port:
+    description:
+      - Port of access from API
+    type: str
+    required: false
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:124]:
```suggestion
    required: false
    version_added: '1.2.0'
```

File: plugins/modules/rabbitmq_user.py [hunk 10]
```diff
@@ -106,6 +106,33 @@
...
+  login_user:
+    description:
+      - User from administrator of the management API
+    type: str
+    required: false
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:129]:
```suggestion
    required: false
    version_added: '1.2.0'
```

File: plugins/modules/rabbitmq_user.py [hunk 11]
```diff
@@ -106,6 +106,33 @@
...
+  login_password:
+    description:
+      - Password from administrator of the management API
+    type: str
+    required: false
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:134]:
```suggestion
    required: false
    version_added: '1.2.0'
```

**Review by Andersson007 — COMMENTED** [review]
@rayhihi hello, thank you for the PR!

I see this is a draft but in the future you'll also need:
- Fix sanity tests (you can run them locally before pushing, see the [Quick start guide](https://docs.ansible.com/ansible/devel/community/create_pr_quick_start.html))
- Add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment)
- Add [integration tests](https://docs.ansible.com/ansible/devel/community/collection_contributors/collection_integration_tests.html)
- Be sure the PR doesn't introduce breaking changes, i.e. users won't get their playbooks broken after installing the collection containing the changes

Thank you!

**Comment by Andersson007** [comment]
@rayhihi please fix the CI tests, if you click Details in the section below in the items marked with red crosses, you'll see the reasons.
Also not all the things listed in https://github.com/ansible-collections/community.rabbitmq/pull/120#pullrequestreview-964201504 are done.

**Comment by rayhihi** [comment]
> @rayhihi please fix the CI tests, if you click Details in the section below in the items marked with red crosses, you'll see the reasons. Also not all the things listed in [#120 (review)](https://github.com/ansible-collections/community.rabbitmq/pull/120#pullrequestreview-964201504) are done.
Oh, I'm sorry I clicked before i finished my repairs, but i'm still working the feature.

File: plugins/modules/rabbitmq_user.py [hunk 12]
```diff
@@ -103,9 +104,38 @@
...
-    required: false
     default: on_create
     choices: ['on_create', 'always']
+  login_protocol:
+    description:
+      - Specify which TCP/IP protocol will be used.
+    type: str
+    default: http
+    choices: ['http', 'https']
+    aliases: [login_protocol, protocol]
```
Comment (arloc) [inline plugins/modules/rabbitmq_user.py:115]:
```suggestion
    aliases: [protocol]
```

File: plugins/modules/rabbitmq_user.py [hunk 13]
```diff
@@ -103,9 +104,38 @@
...
+    version_added: '1.2.0'
+  login_host:
+    description:
+      - Hostname of API.
+    type: str
+    aliases: [login_host, host]
```
Comment (arloc) [inline plugins/modules/rabbitmq_user.py:121]:
```suggestion
    aliases: [host]
```

File: plugins/modules/rabbitmq_user.py [hunk 14]
```diff
@@ -103,9 +104,38 @@
...
+    version_added: '1.2.0'
+  login_port:
+    description:
+      - login_port of access from API.
+    type: str
+    aliases: [login_port, login_port]
```
Comment (arloc) [inline plugins/modules/rabbitmq_user.py:127]:
```suggestion
    aliases: [port]
```

File: plugins/modules/rabbitmq_user.py [hunk 15]
```diff
@@ -461,7 +838,13 @@ def main():
         force=dict(default='no', type='bool'),
         state=dict(default='present', choices=['present', 'absent']),
         node=dict(default='rabbit'),
-        update_password=dict(default='on_create', choices=['on_create', 'always'], no_log=False)
+        update_password=dict(default='on_create', choices=['on_create', 'always'], no_log=False),
+        login_protocol=dict(type="str", default="http", choices=["http", "https"],
+                            aliases=['login_protocol', 'protocol'], no_log=False),
```
Comment (arloc) [inline plugins/modules/rabbitmq_user.py:843]:
```suggestion
                            aliases=['protocol'], no_log=False),
```

File: plugins/modules/rabbitmq_user.py [hunk 16]
```diff
@@ -461,7 +838,13 @@ def main():
...
+        login_host=dict(type="str", aliases=['login_host', 'host'], default=None),
```
Comment (arloc) [inline plugins/modules/rabbitmq_user.py:844]:
```suggestion
        login_host=dict(type="str", aliases=['host'], default=None),
```

File: plugins/modules/rabbitmq_user.py [hunk 17]
```diff
@@ -461,7 +838,13 @@ def main():
...
+        login_port=dict(type="str", aliases=['login_port', 'login_port'], default=None),
```
Comment (arloc) [inline plugins/modules/rabbitmq_user.py:845]:
```suggestion
        login_port=dict(type="str", aliases=['port'], default=None),
```

**Review by arloc — CHANGES_REQUESTED** [review]

File: plugins/modules/rabbitmq_user.py (hunk 12 above)
Comment (Andersson007) [reply plugins/modules/rabbitmq_user.py:115]: @arloc thanks for pointing it out! Yes, we should avoid aliases in general and especially at the initial stage.

Comment (rayhihi) [reply plugins/modules/rabbitmq_user.py]: All done and fixed. =]

File: plugins/modules/rabbitmq_user.py [hunk 18]
```diff
@@ -461,7 +838,13 @@ def main():
         force=dict(default='no', type='bool'),
         state=dict(default='present', choices=['present', 'absent']),
         node=dict(default='rabbit'),
-        update_password=dict(default='on_create', choices=['on_create', 'always'], no_log=False)
+        update_password=dict(default='on_create', choices=['on_create', 'always'], no_log=False),
+        login_protocol=dict(type="str", default="http", choices=["http", "https"],
+                            aliases=['protocol'], no_log=False),
+        login_host=dict(type="str", aliases=['host'], default=None),
+        login_port=dict(type="str", aliases=['port'], default=None),
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:845]:
```suggestion
        login_protocol=dict(type="str", default="http", choices=["http", "https"]),
        login_host=dict(type="str"),
        login_port=dict(type="str"),
```

File: plugins/modules/rabbitmq_user.py [hunk 19]
```diff
@@ -461,7 +838,13 @@ def main():
...
+        login_user=dict(type="str", default=None, no_log=True),
+        login_password=dict(type="str", default=None, no_log=True)
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:847]:
```suggestion
        login_user=dict(type="str", no_log=True),
        login_password=dict(type="str", no_log=True)
```
as far as I remember it's `None` by default for str parameters

File: plugins/modules/rabbitmq_user.py [hunk 20]
```diff
@@ -103,9 +104,38 @@
...
+    type: str
+    default: http
+    choices: ['http', 'https']
+    aliases: [protocol]
+    version_added: '1.2.0'
+  login_host:
+    description:
+      - Hostname of API.
+    type: str
+    aliases: [host]
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:121]:
```suggestion
```

File: plugins/modules/rabbitmq_user.py [hunk 21]
```diff
@@ -103,9 +104,38 @@
...
-    required: false
     default: on_create
     choices: ['on_create', 'always']
+  login_protocol:
+    description:
+      - Specify which TCP/IP protocol will be used.
+    type: str
+    default: http
+    choices: ['http', 'https']
+    aliases: [protocol]
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:115]:
```suggestion
```

File: plugins/modules/rabbitmq_user.py [hunk 22]
```diff
@@ -103,9 +104,38 @@
...
+    description:
+      - Hostname of API.
+    type: str
+    aliases: [host]
+    version_added: '1.2.0'
+  login_port:
+    description:
+      - login_port of access from API.
+    type: str
+    aliases: [login_port]
```
Comment (Andersson007) [inline plugins/modules/rabbitmq_user.py:127]:
```suggestion
```

Comment (rayhihi) [reply plugins/modules/rabbitmq_user.py:127]: I'll work on your suggestions =]

**Comment by nrukavkov** [comment]
When are u gonna merge it? =)

File: plugins/modules/rabbitmq_user.py (hunk 22 above)
Comment (rayhihi) [reply plugins/modules/rabbitmq_user.py:127]: @nrukavkov I need to work on this suggestions.

**Comment by seocam** [comment]
@Andersson007 I've fixed the CI errors and also added the changelog fragment.

Now I'm working on the integration tests but I'm having issues running them locally (I'm following the instructions in the link you have provided). 

That's the command I'm running:
```
ansible-test integration rabbitmq_user --docker -v 
```

That's the output:
```
Falling back to tests in "tests/integration/targets/" because "roles/test/" was not found.
Assuming Docker is available on localhost.
Run command: docker -v
Detected "docker" container runtime version: Docker version 20.10.12, build 20.10.12-0ubuntu4
Starting new "ansible-test-controller-3stxPaop" container.
Run command: docker image inspect quay.io/ansible/default-test-container:5.9.0
Run command: docker run --volume /sys/fs/cgroup:/sys/fs/cgroup:ro --privileged=false --sec ...
Adding "ansible-test-controller-3stxPaop" to container database.
Run command: docker container inspect d43ce50c06cd8e80c6016ef182e06ae98b8db903de7d8975a859 ...
Stream command with data: docker exec -i ansible-test-controller-3stxPaop /bin/sh
Error response from daemon: Container d43ce50c06cd8e80c6016ef182e06ae98b8db903de7d8975a859d35e7d30eae2 is not running
ERROR: Host <ansible_test._internal.host_profiles.DockerProfile object at 0x7f21ee297850> job failed: Command "docker exec -i ansible-test-controller-3stxPaop /bin/sh" returned exit status 1.
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/provisioning.py", line 190, in dispatch_jobs
    thread.wait_for_result()
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/thread.py", line 44, in wait_for_result
    raise exception[1].with_traceback(exception[2])
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/thread.py", line 31, in run
    self._result.put((self.action(), None))
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/provisioning.py", line 131, in provision
    profile.setup()
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/host_profiles.py", line 373, in setup
    docker_exec(self.args, self.container_name, [shell], data=setup_sh, capture=False)
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/docker_util.py", line 529, in docker_exec
    return docker_command(args, ['exec'] + options + [container_id] + cmd, capture=capture, stdin=stdin, stdout=stdout, interactive=interactive,
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/docker_util.py", line 563, in docker_command
    return run_command(args, command + cmd, env=env, capture=capture, stdin=stdin, stdout=stdout, interactive=interactive, always=always,
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/util_common.py", line 420, in run_command
    return raw_command(cmd, capture=capture, env=env, data=data, cwd=cwd, explain=explain, stdin=stdin, stdout=stdout, interactive=interactive,
  File "/home/seocam/ansible_collections/community/rabbitmq/venv/lib/python3.10/site-packages/ansible_test/_internal/util.py", line 492, in raw_command
    raise SubprocessError(cmd, status, stdout_text, stderr_text, runtime, error_callback)

FATAL: Host job(s) failed. See previous error(s) for details.
Run command: docker rm -f ansible-test-controller-3stxPaop
```

I've also tried to look for help in the #ansible channel on Freenode but I got zero responses.

**Comment by csmart** [comment]
> If I try to run the controller container in foreground, just replacing the `-d` by `-ti`, that's what I get:
> 
> ```
> ❯ docker run --volume /sys/fs/cgroup:/sys/fs/cgroup:ro --privileged=false --security-opt seccomp=unconfined --volume /var/run/docker.sock:/var/run/docker.sock --name ansible-test-controller-lI0sqAAl -ti --ulimit nofile=10240 quay.io/ansible/default-test-container:5.9.0
> systemd 245.4-4ubuntu3.15 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid)
> Detected virtualization docker.
> Detected architecture x86-64.
> 
> Welcome to Ubuntu 20.04.3 LTS!
> 
> Set hostname to <e7dde5d863b6>.
> Failed to create /init.scope control group: Read-only file system
> Failed to allocate manager object: Read-only file system
> [!!!!!!] Failed to allocate manager object.
> Exiting PID 1...
> ```

Hi @seocam! I think the problem might be that the host you're running on is configured to use `cgroupsv2` while the way we're running it expects `cgroupsv1`.

I did a test and I can reproduce what you're seeing with `docker` on Fedora 36 with `cgroupsv2`:

```bash
$ docker --debug run -it --volume /sys/fs/cgroup:/sys/fs/cgroup:ro --privileged=false --security-opt seccomp=unconfined quay.io/ansible/default-test-container:5.9.0
systemd 245.4-4ubuntu3.15 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid)
Detected virtualization docker.
Detected architecture x86-64.

Welcome to Ubuntu 20.04.3 LTS!

Set hostname to <cd78a59bdc7e>.
Failed to create /init.scope control group: Read-only file system
Failed to allocate manager object: Read-only file system
[!!!!!!] Failed to allocate manager object.
Exiting PID 1...
DEBU[0001] [hijack] End of stdout                       
```

However, if I set my machine to use `cgroupsv1` instead (https://blog.christophersmart.com/2019/12/15/enabling-docker-in-fedora-31-by-reverting-to-cgroups-v1/) then I can run the container just fine:

```bash
$ docker --debug run -it --volume /sys/fs/cgroup:/sys/fs/cgroup:ro --privileged=false --security-opt seccomp=unconfined quay.io/ansible/default-test-container:5.9.0
systemd 245.4-4ubuntu3.15 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid)
Detected virtualization docker.
Detected architecture x86-64.

Welcome to Ubuntu 20.04.3 LTS!

Set hostname to <60187bc8b750>.
[  OK  ] Created slice system-getty.slice.
...
[  OK  ] Finished Update UTMP about System Runlevel Changes.

Ubuntu 20.04.3 LTS 60187bc8b750 console

60187bc8b750 login: 
```

And then I am also able to use `docker` to run the `ansible-test` successfully:

```bash
$ ansible-test integration rabbitmq_user --docker
Starting new "ansible-test-controller-OF14KbSK" container.
Adding "ansible-test-controller-OF14KbSK" to container database.
Running rabbitmq_user integration test role
[WARNING]: running playbook inside collection community.rabbitmq

PLAY [testhost] ******************************************************************************

TASK [Gathering Facts] ***********************************************************************
ok: [testhost]

TASK [setup_tls : ensure target directory is present] ****************************************
changed: [testhost]

...

PLAY RECAP ***********************************************************************************
testhost                   : ok=55   changed=23   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

I think it should be safe to try switching your machine over to `cgroupsv1`, however if not then perhaps you could try with a VM, or switch to `podman` with the docker shim?

Hope that helps, please let us know how you go!

**Comment by seocam** [comment]
Moving from cgroupsv2 to cgroupsv1 did the trick! I'm working on the unittests

**Comment by csmart** [comment]
> Moving from cgroupsv2 to cgroupsv1 did the trick! I'm working on the unittests

Great! I'm glad it helped. I have created a docs PR upstream to make this more clear for new contributors, so thanks for letting us know :+1:  https://github.com/ansible/ansible/pull/79066

**Comment by seocam** [comment]
Tests added. I'm considering the new version would be 1.3.0, so I've updated the docs and bumped galaxy.yaml.
After doing a rebase with the upstream/main some tests started to fail but they seem to be unrelated with this PR.

**Review by Im0 — CHANGES_REQUESTED** [review]
I believe that the next version will be v1.2.3.  Do you mind changing back to `1.2.3` and updating the version_added?

Otherwise looks good to me.

**Comment by nrukavkov** [comment]
Is there any chance to be released?

**Comment by nrukavkov** [comment]
https://github.com/ansible-collections/community.rabbitmq/pull/155

**Comment by csmart** [comment]
closed in favour of !155

**Pull request closed without merging**